### ISSUE TITLE
kiali header strategy not working as expected

### ISSUE THREAD

**Issue by lake-of-dreams** [kiali/kiali#4464]
Hi, we integrated kiali with keycloak and reverse proxy that we have using the [header strategy implementation guide](https://kiali.io/docs/configuration/authentication/header/). We made the ingress pointing to kiali UI to go through our reverse proxy where we validate the `Authorization` header and then redirect the request to `kiali-server` after adding proper `impersonation` headers as mentioned in the guide.
But it was observed that every time we open kiali, it shows a `Login to Kiali` screen momentarily and sends an `/kiali/api/authenticate` request with an `Authorization` header that contains `username`, `password` and `token` as blank. Now since our reverse proxy expects valid values for the header, we had to add logic to handle this special case but that does not seem like a cleaner approach.
We looked at kiali code and found that the `kiali-ui` sends this request for both `anonymous` and `header` strategy ([link](https://github.com/kiali/kiali-ui/blob/master/src/app/AuthenticationController.tsx#L114)) and there was no way to bypass that.

In our opinion, since it is the reverse proxy which is responsible for Authn and Authz in case of `header` strategy, showing the `Login to Kiali` screen and sending the `/api/authenticate` request from Kiali UI seems an unnecessary step.

Can you please help us mitigate this?

**Comment by lucasponce** [comment]
Hi @lake-of-dreams, thanks for open this issue.

These integrations tend to vary from user to user.
We are happy if you'd like to propose some PR that we can review, our @israel-hdez colleague is familiar with the code but if users would like to add changes, happy to discuss with you on it and you may participate as well in the code.

**Comment by israel-hdez** [comment]
The author of the `header` strategy added the call to `/api/authenticate` to set some cookies that Kiali expects. So, the request to `/api/authenticate` is required for Kiali to work properly.

The `header` strategy was donated by @mlbiam who is a community user. Thus, we expect that community gives maintenance. I think it is possible to remove the request to `/api/authenticate`, or minimally to prevent sending the empty `Authorization` header. I can help with project on-boarding if you are willing to contribute the fix.

Showing the "Login to Kiali" screen with an actionable button is a buggy behavior that I can fix.

**Comment by mlbiam** [comment]
From what I remember the call to `/kiali/authenticate` is what triggers the initiation of the session with a subject access review (since the token can be anything, not just oidc). I don't see what the issue is, as your reverse proxy needs to provide a token regardless of the page. What special logic are you using?

**Comment by jshaughn** [comment]
@lake-of-dreams , I think this is waiting on your response...

**Comment by wmhopkins** [comment]
The issue is that our reverse proxy is expecting to authenticate incoming requests. Using the header strategy with impersonation, we authenticate our user via OIDC, then expect to impersonate that user to Kiali (and Kubernetes). Using the header strategy, we send a service account token with impersonate permission, and the kubernetes impersonation headers for the user we authenticated.

When Kiali subsequently makes a request to the external ingress URL, with the authentication header set, it also passes through our reverse proxy. The proxy sees the Authentication header and tries to authenticate a user based on the supplied credentials. Since there aren't actually any valid credentials in the header, authentication fails and the request is rejected (without the temporary workaround to whitelist the `/kiali/authenticate` URI).

This seems like it would be a problem for almost any reverse proxy sitting in front of Kiali. If the proxy expects to authenticate users, an incoming request that purports to contain credentials (i.e., an Authorization header), but in fact contains invalid credentials, or credentials not understood by the proxy, should normally be rejected.

**Comment by wmhopkins** [comment]
To put it a bit more succinctly, the problem isn't that Kiali is making a call to `/kiali/authenticate', it's that it's setting an invalid `Authorization` header first. Arguably, it shouldn't be setting the `Authorization` header at all if it's expecting a reverse proxy to provide credentials; i.e., if it's in header authentication mode.

**Comment by mlbiam** [comment]
> To put it a bit more succinctly, the problem isn't that Kiali is making a call to /kiali/authenticate, it's that it's setting an invalid Authorization header first. Arguably, it shouldn't be setting the Authorization header at all if it's expecting a reverse proxy to provide credentials; i.e., if it's in header authentication mode.

While it does look like kiali-ui is setting the `Authorization` header to an empty username and password (https://github.com/kiali/kiali-ui/blob/03e8b8cf276782b05233eedde43e9ac8d4b94f54/src/services/Api.ts#L93), your reverse proxy should be stripping out any `Authorization` header anyways for security reasons (as well as making sure there are no impersonation headers too) regardless of what comes from the browser.

kiali-ui probably shouldn't be sending the `Authorization` header, but i don't think its a critical issue since it's the responsibility of the reverse proxy to clean out incoming headers (such as `Authorization`) to avoid credential leaks or escalation attacks.

**Comment by wmhopkins** [comment]
> While it does look like kiali-ui is setting the `Authorization` header to an empty username and password (https://github.com/kiali/kiali-ui/blob/03e8b8cf276782b05233eedde43e9ac8d4b94f54/src/services/Api.ts#L93), your reverse proxy should be stripping out any `Authorization` header anyways for security reasons (as well as making sure there are no impersonation headers too) regardless of what comes from the browser.

Our proxy supports authentication with both bearer tokens and basic auth credentials, so we can't just strip out `Authorization` headers. If we see an `Authorization` header, we have to assume that the client intended to authenticate the request using those credentials.

Conversely, I can't think of a good reason to attach an invalid `Authorization` header to a request, nor does it seem reasonable to expect third parties to understand that the header doesn't contain real credentials and should be ignored.

**Comment by mlbiam** [comment]
> Our proxy supports authentication with both bearer tokens and basic auth credentials, so we can't just strip out Authorization headers. If we see an Authorization header, we have to assume that the client intended to authenticate the request using those credentials.

I'd say that's a design flaw and a security risk.  If you already support OIDC to the proxy and are also supporting basic and bearer tokens to the same URL?  Smells like you're leaving yourself open to some kind of session hijack.  You should be able to limit which URLs accept specific kinds of tokens.  Kiali is a web app so you should be using a session cookie for managing the session with the reverse proxy, not allowing someone to override it with a token.  

> Conversely, I can't think of a good reason to attach an invalid Authorization header to a request, nor does it seem reasonable to expect third parties to understand that the header doesn't contain real credentials and should be ignored.

Yes, that would be why I said it's a bug.  Just not a critical one.  That said, it's a great /good-first-issue already pointed out where the code that needs changing is.

**Comment by wmhopkins** [comment]
> > Our proxy supports authentication with both bearer tokens and basic auth credentials, so we can't just strip out Authorization headers. If we see an Authorization header, we have to assume that the client intended to authenticate the request using those credentials.
> 
> I'd say that's a design flaw and a security risk. If you already support OIDC to the proxy and are also supporting basic and bearer tokens to the same URL? Smells like you're leaving yourself open to some kind of session hijack. You should be able to limit which URLs accept specific kinds of tokens. Kiali is a web app so you should be using a session cookie for managing the session with the reverse proxy, not allowing someone to override it with a token.

With all due respect, since you don't know what problems we're trying to solve, you have no basis for evaluating our design. We have use cases for both types of authentication, and it's not clear to me that handling bearer tokens or basic auth is intrinsically a "security risk".

Because we also manage session cookies, if a request has both a cookie and an `Authorization` header, we discard the existing cookie and re-authenticate using the supplied credentials. To throw away the header and rely on the cookie seems backwards, and is almost certainly not what the client intended. Failing a request that presents bad credentials seems more secure than simply discarding them, whether an `Authorization` header is expected or not.
 
> > Conversely, I can't think of a good reason to attach an invalid Authorization header to a request, nor does it seem reasonable to expect third parties to understand that the header doesn't contain real credentials and should be ignored.
> 
> Yes, that would be why I said it's a bug. Just not a critical one. That said, it's a great /good-first-issue already pointed out where the code that needs changing is.

OK. Critical is in the eye of the beholder, I guess. I'd be happy to fix it, but unfortunately I can't contribute to an open source project without getting approval from various Oracle powers-that-be. I'll look into it. Meanwhile we'll special-case the `/kiali/authenticate` URI and/or fix the problem in our build-from-source Kiali image. Thanks for pointing out the code that needs fixing.

**Comment by mlbiam** [comment]
think it is resolved, so much as we've identified the issue and where it needs to be fixed.  i'm spread thin right now so i can't do it at the moment. If someone else isn't able to tackle it by the time things calm down for me i'll be happy to fix it.

**Comment by wmhopkins** [comment]
Agree, the issue is understood. We have workarounds for now, and we'll look at getting approval to contribute a fix if nobody beats us to it.

**Comment by Hronom** [comment]
Actually same issue here. 
Try to use `oauth2-proxy` with Kiali configured with `header` strategy and you be able to reproduce the issue.

`oauth2-proxy` config:
```
        oauth2-proxy:
          customLabels:
            sidecar.istio.io/inject: 'true'
          fullnameOverride: kiali-oauth2-proxy
          config:
            existingSecret: blablabla
            configFile: |-
              provider = "oidc"
              provider_display_name = "JumpCloud"
              custom_sign_in_logo = "https://raw.githubusercontent.com/kiali/kiali.io/staging/static/favicons/android-192x192.png"
              banner = "Kiali"
              oidc_issuer_url = "https://oauth.id.jumpcloud.com/"
              upstreams = [
                "http://kiali:20001/"
              ]
              pass_basic_auth = true
              email_domains = [
                "*"
              ]
              pass_authorization_header = true
              pass_access_token = true
              set_authorization_header = true
```

@wmhopkins please if you can share your workaround it will be very helpful!

**Comment by jshaughn** [comment]
@Hronom Any update?

**Comment by Hronom** [comment]
@jshaughn I stopped here, I don't have time for now to go deeper. Please if you are in context and have possibility try `oauth2-proxy` with the config that I posted above. `oauth2-proxy` pretty popular thing.

For now I set Kiali in view only mode and `oauth2-proxy` in front helps to authenticate user before it able to access Kiali.

In general it will be cool to have in Kiali same OIDC/SSO support as in ArgoCD for example, but this is other topic...

**Comment by jshaughn** [comment]
@Hronom, thanks for the update.  At the moment the Kiali maintainers have no bandwidth to expand our auth support, we're hoping someone can help out, otherwise it will be some time before we can do much here.

**Comment by jshaughn** [comment]
This code has been revised many time since the report. It may or may not still be an issue, but closing in lieu of a new issue.
